private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go

```diff
@@ -0,0 +1,244 @@
+) (_ bufmodule.Module, retErr error) {
+	moduleBasedir := normalpath.Join(c.baseDir, modulePin.Remote(), modulePin.Owner(), modulePin.Repository())
+	var manifestHexDigest string
+	if modulePinDigestEncoded := modulePin.Digest(); modulePinDigestEncoded != "" {
+		modulePinDigest, err := manifest.NewDigestFromString(modulePinDigestEncoded)
+		if err != nil {
+			return nil, err
+		}
+		manifestHexDigest = modulePinDigest.Hex()
+	} else {
+		// Attempt to look up manifest digest from commit
+		commitPath := normalpath.Join(moduleBasedir, commitsDir, modulePin.Commit())
+		manifestDigestBytes, err := os.ReadFile(commitPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, storage.NewErrNotExist(commitPath)
+			}
+			return nil, err
+		}
+		manifestHexDigest = strings.TrimSpace(string(manifestDigestBytes))
+		if manifestHexDigest == "" {
+			return nil, storage.NewErrNotExist(commitPath)
+		}
+	}
+	manifestPath := normalpath.Join(moduleBasedir, manifestsDir, manifestHexDigest[:2], manifestHexDigest[2:])
+	f, err := os.Open(manifestPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, storage.NewErrNotExist(manifestPath)
+		}
+		return nil, err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, f.Close())
+	}()
+	cacheManifest, err := manifest.NewFromReader(f)
+	if err != nil {
+		return nil, err
+	}
+	cacheManifestBlob, err := cacheManifest.Blob()
+	if err != nil {
+		return nil, err
+	}
+	if cacheManifestBlob.Digest().Hex() != manifestHexDigest {
+		return nil, storage.NewErrNotExist(fmt.Sprintf("digest mismatch - expected: %q, found: %q", manifestHexDigest, cacheManifestBlob.Digest().Hex()))
+	}
+	var blobs []manifest.Blob
+	blobBasedir := normalpath.Join(moduleBasedir, blobsDir)
+	blobDigests := make(map[string]struct{})
+	for _, path := range cacheManifest.Paths() {
+		digest, found := cacheManifest.DigestFor(path)
+		if !found {
+			return nil, storage.NewErrNotExist(path)
+		}
+		hexDigest := digest.Hex()
+		if _, ok := blobDigests[hexDigest]; ok {
+			// We've already loaded this
+			continue
+		}
+		blobPath := normalpath.Join(blobBasedir, hexDigest[:2], hexDigest[2:])
+		contents, err := os.ReadFile(blobPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, storage.NewErrNotExist(path)
+			}
+			return nil, err
+		}
+		blob, err := manifest.NewMemoryBlob(*digest, contents, manifest.MemoryBlobWithDigestValidation())
+		if err != nil {
+			return nil, err
+		}
+		blobs = append(blobs, blob)
+	}
+	blobSet, err := manifest.NewBlobSet(ctx, blobs)
+	if err != nil {
+		return nil, storage.NewErrNotExist(manifestPath)
+	}
+	bucket, err := manifest.NewBucket(
+		*cacheManifest,
+		*blobSet,
+		manifest.BucketWithAllManifestBlobsValidation(),
+		manifest.BucketWithNoExtraBlobsValidation(),
+	)
+	if err != nil {
+		return nil, storage.NewErrNotExist(manifestPath)
+	}
+	return bufmodule.NewModuleForBucket(ctx, bucket, bufmodule.ModuleWithManifestAndBlobs(*cacheManifest, *blobSet))
```

Comment:
This feels a bit redundant, if we have the bucket we shouldn't need the m+b, and viceversa, no need for the bucket, we're passing the m+b.

Reply:
I thought about removing this, but we at least need the manifest at a higher level to compare the manifest of the cached bucket to the one defined in the `buf.lock` file. Maybe we create a shallow module which has a manifest and no blobs?

Reply:
I see no issue in exposing `Manifest` and `Blobs` getter in the module interface, that way you could compare if it's the same with the cached one.

Reply:
Yea the constructor should not take the manifest

Reply:
Created a new constructor so we no longer have an option for adding a manifest and blobs - we create the module from them.